de.dashup.model/src/main/java/de/dashup/model/service/DashupService.java

```diff
@@ -109,13 +109,161 @@ public Widget getPanelById(int id) throws SQLException {
+        Map<String, Object> whereParameters = new HashMap<>();
+        Map<String, String> onParameters = new HashMap<>();
+        ArrayList<String> returningValue = new ArrayList<>();
+        whereParameters.put("panel_id", widgetId);
+        onParameters.put("tag_id", "id");
+        JSONArray databaseResult = this.database.get(Database.Table.PANELS_TAGS, Database.Table.TAGS, onParameters, whereParameters);
+        for (int i = 0; i < databaseResult.length(); i++) {
+            returningValue.add(databaseResult.getJSONObject(i).getString("text"));
+        }
+        return returningValue;
+    }
+
+    public List<Rating> getRatingsByWidgetID(int widgetId) throws SQLException {
+        ArrayList<Rating> returningValue = new ArrayList<>();
+        Map<String, String> onParameters = new HashMap<>();
+        Map<String, Object> whereParameters = new HashMap<>();
+        onParameters.put("user_id", "id");
+        whereParameters.put("panel_id", widgetId);
+        List<? extends DatabaseObject> databaseResult = database.getObject(Database.Table.RATINGS, Database.Table.USERS, Rating.class, onParameters, whereParameters, "ratings.changed_on DESC");
+        for (DatabaseObject databaseObject : databaseResult) {
+            Rating rating = (Rating) databaseObject;
+            returningValue.add(rating);
+        }
+        return returningValue;
+    }
+
+    public List<Widget> getTopWidgets(String sortByField) throws SQLException {
+        ArrayList<Widget> returningValue = new ArrayList<>();
+        Map<String, Object> whereParameters = new HashMap<>();
+        whereParameters.put("visibility", 1);
+        List<? extends DatabaseObject> result = this.database.getObject(Database.Table.PANELS, DatabaseWidget.class, whereParameters, sortByField + " DESC");
+        if (result != null && result.size() >= 4) {
+            for (int i = 0; i < 4; i++) {
+                Widget widget = new Widget().fromDatabaseObject(result.get(i));
+                widget.setShortDescription(this.shortenShortDescOfPanel(widget.getShortDescription()));
+                returningValue.add(widget);
+            }
+        } else if (result != null) {
+            for (DatabaseObject databaseObject : result) {
+                Widget widget = new Widget().fromDatabaseObject(databaseObject);
+                widget.setShortDescription(this.shortenShortDescOfPanel(widget.getShortDescription()));
+                returningValue.add(widget);
+            }
+        }
+        return returningValue;
+    }
+
+    public Map<Widget, Rating> getFeaturedWidgets(int[] widgetIds) throws SQLException {
+        Map<Widget, Rating> returningValue = new HashMap<>();
+        for (int widgetId : widgetIds) {
+            returningValue.put(this.getPanelById(widgetId), this.getTopRating(widgetId));
+        }
+        return returningValue;
+    }
+
+    private Rating getTopRating(int widgetId) throws SQLException {
```

Comment:
Please use the SQL AVG Aggregation and don't take the best ranking for a widget, but the average one.

Reply:
On the carusel the rating is taken from the widget. But i need this function to get the featured comment so this is absolutely correct. We agreed on a hard coded carusel so here is it... on db we can create comments with rating 101 so they are used for the carusel...

Reply:
Ah, now I understand the use case of this method. Didn't know that this is used for the carousel, thought it is used for marketplace result list ^^. My fault!